unittests/test-contracts/if_ibc/if_ibc.cpp

```diff
@@ -0,0 +1,252 @@
+            if (first){
+                first=false;
+                agg_pub_key = fa_itr->public_key;
+            }
+            else agg_pub_key = _g1add(agg_pub_key, fa_itr->public_key);
+            weight+=fa_itr->fweight;
+
+        }
+        index++;
+        fa_itr++;
+    }
+
+    //verify that we have enough vote weight to meet the quorum threshold of the target policy
+    check(weight>=target_policy.fthreshold, "insufficient signatures to reach quorum");
+
+    std::array<uint8_t, 32> data = finality_digest.extract_as_byte_array();
+    std::vector<const char> v_data(data.begin(), data.end());
+
+    //verify signature validity
+    _verify(agg_pub_key, qc.signature, v_data);
+
+
+}
+
+void if_ibc::_check_target_block_proof_of_inclusion(const proof_of_inclusion& proof, const std::optional<checksum256> reference_root){
+
+    //verify that the proof of inclusion is over a target block
+    check(std::holds_alternative<if_ibc::block_data>(proof.target), "must supply proof of inclusion over block data");
+
+    //resolve the proof to its merkle root
+    checksum256 finality_mroot = proof.root();
+
+    if (reference_root.has_value()){
+        check(reference_root.value() == finality_mroot, "cannot link proof to proven merkle root");
+    }
+    else {
+
+        proofs_table _proofs_table(get_self(), get_self().value);
+
+        auto merkle_index = _proofs_table.get_index<"merkleroot"_n>();
+
+        auto itr = merkle_index.find(finality_mroot);
+        check(itr!= merkle_index.end(), "cannot link proof to proven merkle root");
+
+    }
+
+    block_data target_block = std::get<if_ibc::block_data>(proof.target);
+
+    if (target_block.finality_data.active_finalizer_policy.has_value()){
+        _maybe_set_finalizer_policy(target_block.finality_data.active_finalizer_policy.value(), target_block.dynamic_data.block_num);
+    }
+    
+}
+
+void if_ibc::_check_finality_proof(const finality_proof& finality_proof, const proof_of_inclusion& target_block_proof_of_inclusion){
+
+    //temporarilly disabled : skip qc verification
+    //if QC is valid, it means that we have reaced finality on the block referenced by the finality_mroot
+    //_check_qc(finality_proof.qc, finality_proof.qc_block.finality_digest(), finality_proof.qc_block.finalizer_policy_generation);
+
+    //check if the target proof of inclusion correctly resolves to the root of the finality proof
+    _check_target_block_proof_of_inclusion(target_block_proof_of_inclusion, finality_proof.qc_block.finality_mroot);
+    
+    //if proof of inclusion was successful, the target block and its dynamic data have been validated as final and correct
+    block_data target_block = std::get<if_ibc::block_data>(target_block_proof_of_inclusion.target);
+
+    //if the finality_mroot we just proven is more recent than the last root we have stored, store it
+    _maybe_add_proven_root(target_block.dynamic_data.block_num, finality_proof.qc_block.finality_mroot);
+
+}

```

Comment:
I don't see base_digest being used.

Reply:
The base_digest is actually called witness_hash here, and its content will depend on whether or not was have an new finality policy. You can see here :

https://github.com/AntelopeIO/leap/pull/2314/files#diff-dbff99e45bcfb34c437a7a084bd325c478280c758e3282913923e42ce7afba6aR264

If you think this is preferrable, I can rename it for clarity.

Reply:
No need to rename. Probably add a comment.